zne/noise_amplification/folding_amplifier/glorious_folding_amplifier.py

```diff
@@ -0,0 +1,100 @@
+    foldings, and :math:`s` is the number of gates that are sub-folded [1].
+
+    The gates of the full circuit that are used for the sub-folding can be the first :math:`s`
+    gates, the last :math:`s` gates or a random selection of :math:`s` gates. This is specified by
+    ``sub_folding_option``.
+
+    Note that all instance objects and their attributes are immutable by construction. If another
+    strategy with different options needs to be utilized, a new instance of the class should be
+    created.
+
+    References:
+        [1] T. Giurgica-Tiron et al. (2020).
+            Digital zero noise extrapolation for quantum error mitigation.
+            `<https://ieeexplore.ieee.org/document/9259940>`
+    """
+
+    def amplify_dag_noise(self, dag: DAGCircuit, noise_factor: float) -> DAGCircuit:
+        self._validate_noise_factor(noise_factor)
+        barrier = Barrier(dag.num_qubits())
+        num_full_foldings = self._compute_folding_nums(noise_factor)
+        noisy_dag = copy.deepcopy(dag)
+        noisy_dag = self._apply_full_folding(noisy_dag, dag, barrier, num_full_foldings)
+        return noisy_dag
+
+    def _apply_full_folding(
+        self,
+        noisy_dag: DAGCircuit,
+        dag: DAGCircuit,
+        barrier: Barrier,
+        num_foldings: int,
+    ) -> DAGCircuit:
+        """Fully folds the original DAG circuit a number of ``num_foldings`` times.
+
+        Args:
+            noisy_circuit: The noise amplified circuit to which the gates are added.
+            original_circuit: The original circuit without foldings.
+            num_foldings: Number of times the circuit should be folded.
+
+        Returns:
+            The noise amplified circuit.
+        """
+        inverse_dag = self._invert_dag(dag)
+        for _ in range(num_foldings):
+            noisy_dag.apply_operation_back(barrier, qargs=noisy_dag.qubits)
+            noisy_dag.compose(inverse_dag, inplace=True)
+            noisy_dag.apply_operation_back(barrier, qargs=noisy_dag.qubits)
+            noisy_dag.compose(dag, inplace=True)
+        # noisy_dag.apply_operation_back(barrier, qargs=noisy_dag.qubits)
+        return noisy_dag
+
+    def _invert_dag(self, dag_to_inverse: DAGCircuit) -> DAGCircuit:
+        """Inverts dag circuit"""
+        dag_to_inverse = dag_to_circuit(dag_to_inverse).inverse()
+        return circuit_to_dag(dag_to_inverse)
+
+    def _validate_noise_factor(self, noise_factor: float) -> None:
+        """Validates noise factor."""
+        if noise_factor < 1:
+            raise ValueError(
+                f"{self.name} expects a positive float noise_factor >= 1."
+                f"Received {noise_factor} instead."
+            )
+        if noise_factor % 2 == 0:
+            raise ValueError(
+                f"{self.name} expects a positive odd noise_factor."
+                f"Received {noise_factor} instead."
+            )
```

Comment:
Nice!! 

What if the user provides `noise_factor="3"`? Maybe you want to validate type as well (i.e. raise `TypeError`s), or even better, normalize (i.e. convert the input to a standard type)!

If you decide to follow the latter approach, then all _`validate`_ functions should `return` the normalized input (or the input itself if no normalization is performed). This is in fact the standard behavior that I have adopted now and I like it quite a bit!
```python
noise_factor = "3"
noise_factor = self._validate_noise_factor(noise_factor)
assert noise_factor == 3
assert isinstance(noise_factor, float)
```
There is a `[+1, -1]` lines change (i.e. delete one line, add one line) to accomplish this behavior, can you solve the puzzle? 😉